Sources/NativeBuilder/ContainerBuildSnapshotter/TarArchiveDiffer.swift

```diff
@@ -0,0 +1,649 @@
+                guard let pathString = path.stringValue else {
+                    // Skip files with non-UTF8 paths that can't be represented on this system
+                    continue
+                }
+                let sourceFile = sourceDirectory.appendingPathComponent(pathString)
+                if FileManager.default.fileExists(atPath: sourceFile.path) {
+                    // Prefer Diff payload for uid/gid/permissions overrides
+                    entriesToArchive.append(
+                        Archiver.ArchiveEntryInfo(
+                            pathOnHost: sourceFile,
+                            pathInArchive: URL(fileURLWithPath: pathString),
+                            owner: a.uid,
+                            group: a.gid,
+                            permissions: a.permissions?.rawValue
+                        )
+                    )
+
+                    if inspectorOptions.enableXAttrsCapture {
+                        // Prefer xattrs from Diff payload; fall back to inspector if absent
+                        let payloadXAttrs: [XAttrEntry]? = {
+                            guard let dict = a.xattrs, !dict.isEmpty else { return nil }
+                            let keys = Array(dict.keys).sorted()
+                            return keys.map { key in XAttrEntry(key: key, value: dict[key]!) }
+                        }()
+
+                        let entries: [XAttrEntry]
+                        if let payload = payloadXAttrs {
+                            entries = payload
+                        } else if let fromInspector = try? await inspector.listXAttrs(url: sourceFile, options: inspectorOptions),
+                            !fromInspector.isEmpty
+                        {
+                            entries = fromInspector
+                        } else {
+                            entries = []
+                        }
+
+                        if !entries.isEmpty {
+                            let blob = XAttrCodec.encode(entries)
+                            let sidecarRel = ".container/xattrs/\(pathString).bin"
+                            let sidecarHost = tempDir.appendingPathComponent(sidecarRel)
+
+                            try FileManager.default.createDirectory(
+                                at: sidecarHost.deletingLastPathComponent(),
+                                withIntermediateDirectories: true
+                            )
+                            try blob.write(to: sidecarHost)
+
+                            entriesToArchive.append(
+                                Archiver.ArchiveEntryInfo(
+                                    pathOnHost: sidecarHost,
+                                    pathInArchive: URL(fileURLWithPath: sidecarRel)
+                                )
+                            )
+                        }
+                    }
+                }
+
+            case .modified(let m):
```

Comment:
Is this doing the same thing as the add case?

Reply:
Great question. Even both branches eventually emit an `entry`, `.add` handles paths absent in the baseline, whereas `.modified` applies changes to an existing path with different preconditions and bookkeeping. Keeping them separate keeps those invariants explicit.